sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,140 @@
+    cdef double result, tmp
+    result = 0
+    cdef int i
+    for i in range(n_features):
+        tmp = (a[0] - b[0])
+        a += 1
+        b += 1
+        result += tmp * tmp
+    return sqrt(result)
+
+
+cdef assign_labels(double* X, double* centers, double[:, :]
+                   center_distances, int[:] labels, double[:] distances, int n_samples, int n_features, int n_clusters):
+    print("start assign labels")
+    # assigns closest center to X
+    # uses triangle inequality
+    cdef double* x
+    cdef double* c
+    cdef double d_c, dist
+    cdef int c_x, j, sample
+    for sample in range(n_samples):
+        # assign first cluster center
+        c_x = 0
+        x = X + sample * n_features
+        d_c = d(x, centers, n_features)
+        for j in range(n_clusters):
+            if d_c > center_distances[c_x, j]:
+                c = centers + j * n_features
+                dist = d(x, c, n_features)
+                if dist < d_c:
+                    d_c = dist
+                    c_x = j
+        labels[sample] = c_x
+        distances[sample] = d_c
+    print("end assign labels")
+
+
+def k_means_elkan(X_, int n_clusters, init, float tol=1e-4, int max_iter=30, verbose=False):
+    #initialize
+    tol = _tolerance(X_, tol)
+    centers_ = init
+    cdef double* centers_p = getfloatpointer(centers_)
+    cdef double* X_p = getfloatpointer(X_)
+    cdef double* x_p
+    cdef int n_samples = X_.shape[0]
+    cdef int n_features = X_.shape[1]
+    cdef int point_index, center_index, label
+    cdef float upper_bound, distance
+    cdef double[:, :] center_distances = euclidean_distances(centers_) / 2.
+    cdef double[:, :] lower_bounds = np.zeros((n_samples, n_clusters))
+    cdef double[:] distance_next_center
+    labels_ = np.empty(n_samples, dtype=np.int32)
+    cdef int[:] labels = labels_
+    upper_bounds_ = np.empty(n_samples, dtype=np.float)
+    cdef double[:] upper_bounds = upper_bounds_
+    assign_labels(X_p, centers_p, center_distances, labels, upper_bounds, n_samples, n_features, n_clusters)
+    # make bounds tight for current labelss
+    for point_index in range(n_samples):
+        lower_bounds[point_index, labels[point_index]] = upper_bounds[point_index]
+    cdef np.uint8_t[:] bounds_tight = np.ones(n_samples, dtype=np.uint8)
+    cdef np.uint8_t[:] points_to_update = np.zeros(n_samples, dtype=np.uint8)
+    for iteration in range(max_iter):
+        print("start iteration")
+        # we could find the closest center in O(n) but
+        # this does not seem to be the bottleneck
+        distance_next_center = np.sort(center_distances, axis=0)[1]
+        print("done sorting")
+        for point_index in range(n_samples):
+            upper_bound = upper_bounds[point_index]
+            label = labels[point_index]
+            if distance_next_center[label] >= upper_bound:
+                continue
+            x_p = X_p + point_index * n_features
```

Comment:
Good point, though that is slightly non-trivial because of my awkward mixture of memory views and numpy arrays :-/